playbooks/files/loki/grafana_provisioning.yml.j2

```
@@ -0,0 +1,10 @@
+apiVersion: 1
+
+datasources:
+  - name: Loki
+    type: loki
+    access: proxy
+    url: http://localhost:{{ loki__http_listen_port }}
+    jsonData:
+      timeout: 60
+      maxLines: 1000
```

The loki URL is a site configuration, so it should not be hardcoded to `localhost`. Instead, it should be defined in the inventory groups/hosts variables.
Also, to support a more flexible configuration, I'd place the configuration in an array following the configuration structure. for example:

```yaml
grafana__datasources_loki:
  name: Loki
  access: proxy
  url: http://localhost:{{ loki__http_listen_port }}
```

And then, in this template I'd loop on that variable while setting defaults like this:
```suggestion
datasources:
{% for ds in grafana__datasources_loki %}
{% set defaults = {
  'type': 'loki',
  'name': 'Loki',
  'access': 'proxy',
  'jsonData': {
    'timeout': 60,
    'maxLines': 1000,
  }
}
%}
  - {{ defaults | combine(ds) | to_nice_yaml(indent=2) | indent(4) }}
{% endfor %}
```

---

I've tried to apply your suggestion, but with some changes:

I have placed the following variable in grafana role defaults (`roles/grafana/defaults/main.yaml`). We can add this variable in inventory as well if needed. Is that ok? Or we shouldn't have defaults in the role? 
Moreover I have removed `access: proxy` since I think it will always be the same and it is in the defaults and add a '-' to make the loop work:
```
grafana__datasources_loki:
  - name: Loki
    url: http://localhost:3100
```
PTAL.

---

> I have placed the following variable in grafana role defaults (`roles/grafana/defaults/main.yaml`). We can add this variable in inventory as well if needed. Is that ok? Or we shouldn't have defaults in the role?

As I commented, I think it makes sense to leave the default empty, as the role cannot know if you are also running loki, or where it is located. And then in the inventory variables you override it with the proper values.

> Moreover I have removed `access: proxy` since I think it will always be the same and it is in the defaults

From the docs I am not sure which one is the default, so I think it makes sense to leave it explicit, and it can always be overridden by the user

> and add a '-' to make the loop work:

I did not understand this bit, I see no `-` added to my suggestion?

---

> > Moreover I have removed `access: proxy` since I think it will always be the same and it is in the defaults
> 
> From the docs I am not sure which one is the default, so I think it makes sense to leave it explicit, and it can always be overridden by the user

When I said `defaults` I meant our variable `defaults` in `roles/grafana/templates/ds/loki.yaml.j2`, not provisioning defaults. Therefore, I think we agree.

> > and add a '-' to make the loop work:
> 
> I did not understand this bit, I see no `-` added to my suggestion?

Yes, I added a `- ` to `grafana__datasources_loki`. 

Your suggestion was:
```
grafana__datasources_loki:
  name: Loki
  access: proxy
  url: http://localhost:{{ loki__http_listen_port }}
```
and I think a `-` was missing:
```
grafana__datasources_loki:
  - name: Loki
    access: proxy
    url: http://localhost:3100
```
But, anyway, this is not in this PR any more, since `grafana__datasources_loki` is empty now in the role.